File: app/src/main/java/workshop1024/com/xproject/controller/fragment/home/NewsListFragment.java
package workshop1024.com.xproject.controller.fragment.home;

import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import workshop1024.com.xproject.R;
import workshop1024.com.xproject.controller.adapter.BigCardsAdapter;
import workshop1024.com.xproject.controller.adapter.CompactAdapter;
import workshop1024.com.xproject.controller.adapter.MinimalAdapter;
import workshop1024.com.xproject.controller.fragment.SubFragment;
import workshop1024.com.xproject.model.news.News;
import workshop1024.com.xproject.model.news.source.NewsDataSource;
import workshop1024.com.xproject.model.news.source.NewsRepository;
import workshop1024.com.xproject.view.recyclerview.RecyclerViewItemDecoration;

/**
 * 抽屉导航Home Fragment的PageFragment列表选项点击后展示的子列表Fragment
 */
public class NewsListFragment extends SubFragment implements SwipeRefreshLayout.OnRefreshListener,
        NewsDataSource.LoadNewsListCallback {
    private static final String SEARCH_TYPE = "search_Type";
    private static final String SEARCH_CONTENT = "search_Content";

    private View mRootView;
    private SwipeRefreshLayout mSwipeRefreshLayoutPull;
    private RecyclerView mStoryRecyclerView;
    private RecyclerView.Adapter mListAdapter;

    private String mSearchType;
    private String mSearchContent;
    private NewsRepository mNewsRepository;
    private List<News> mNewsList;

    public NewsListFragment() {
    }

    public static NewsListFragment newInstance(String searchType, String searchContent) {
        NewsListFragment fragment = new NewsListFragment();
        fragment.setNavigationItemId(R.id.leftnavigator_menu_home);
        Bundle args = new Bundle();
        args.putString(SEARCH_TYPE, searchType);
        args.putString(SEARCH_CONTENT, searchContent);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mSearchType = getArguments().getString(SEARCH_TYPE);
            mSearchContent = getArguments().getString(SEARCH_CONTENT);

        }
        mNewsRepository = NewsRepository.getInstance();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mRootView = inflater.inflate(R.layout.homelist_fragment, container, false);

        mSwipeRefreshLayoutPull = mRootView.findViewById(R.id.homelist_swiperefreshlayout_pullrefresh);
        mSwipeRefreshLayoutPull.setOnRefreshListener(this);

        mStoryRecyclerView = mRootView.findViewById(R.id.homelist_recyclerview_list);
        mStoryRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mStoryRecyclerView.addItemDecoration(new RecyclerViewItemDecoration(6));

        return mRootView;
    }

    @Override
    public void onStart() {
        super.onStart();
        refreshNewsList();
    }

    @Override
    public void onRefresh() {
        refreshNewsList();
    }

    private void refreshNewsList() {
        Snackbar.make(mRootView, "Fetch more newses ...", Snackbar.LENGTH_SHORT).show();
        mSwipeRefreshLayoutPull.setRefreshing(true);
        if ("Subscribe".equals(mSearchType)) {
            mNewsRepository.getNewsListBySubscribe(mSearchContent, this);
        } else if ("Tag".equals(mSearchType)) {
            mNewsRepository.getNewsListByTag(mSearchContent, this);
        } else if ("Filter".equals(mSearchType)) {
            mNewsRepository.getNewsListByFilter(mSearchContent, this);
        } else if ("Search".equals(mSearchType)) {
            mNewsRepository.getNewsListBySearch(mSearchContent, this);
        }
    }

    @Override
    public void onNewsLoaded(List<News> newsList) {
        mNewsList = newsList;
        mSwipeRefreshLayoutPull.setRefreshing(false);
        showBigCardsList();
        Snackbar.make(mRootView, "Fetch " + newsList.size() + " newses ...", Snackbar.LENGTH_SHORT).show();
    }

    @Override
    public void onDataNotAvaiable() {
        Snackbar.make(mRootView, "No newses refresh...", Snackbar.LENGTH_SHORT).show();
    }

    public void showBigCardsList() {
        mListAdapter = new BigCardsAdapter(getContext(), mNewsList);
        mStoryRecyclerView.setAdapter(mListAdapter);
    }

    public void showMinimalList() {
        mListAdapter = new MinimalAdapter(getContext(), mNewsList);
        mStoryRecyclerView.setAdapter(mListAdapter);
    }

    public void showCompactList() {
        mListAdapter = new CompactAdapter(getContext(), mNewsList);
        mStoryRecyclerView.setAdapter(mListAdapter);
    }

    public void markAsRead() {
        //将现有文章置为可读
        List<String> newsIdList = new ArrayList<>();
        for (News news : mNewsList) {
            newsIdList.add(news.getNewId());
        }
        mNewsRepository.markNewsesReadedByNewsId(newsIdList);

        //刷新列表
        refreshNewsList();
    }
}

File: home/src/main/java/workshop1024/com/xproject/home/controller/fragment/news/NewsListFragment.kt
package workshop1024.com.xproject.home.controller.fragment.news

import android.os.Bundle
import android.util.EventLog
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.snackbar.Snackbar
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import workshop1024.com.xproject.base.controller.fragment.XFragment
import workshop1024.com.xproject.base.view.recyclerview.RecyclerViewItemDecoration
import workshop1024.com.xproject.home.R
import workshop1024.com.xproject.home.controller.adapter.BigCardsAdapter
import workshop1024.com.xproject.home.controller.adapter.CompactAdapter
import workshop1024.com.xproject.home.controller.adapter.MinimalAdapter
import workshop1024.com.xproject.home.controller.event.NewsListAsReadEvent
import workshop1024.com.xproject.home.controller.event.NewsListShowBigCardsEvent
import workshop1024.com.xproject.home.controller.event.NewsListShowCompactEvent
import workshop1024.com.xproject.home.controller.event.NewsListShowMinimalEvent
import workshop1024.com.xproject.home.databinding.NewslistFragmentBinding
import workshop1024.com.xproject.home.model.Injection
import workshop1024.com.xproject.home.model.news.News
import workshop1024.com.xproject.home.model.news.source.NewsDataSource

/**
 * 新闻列表Fragmnet，用于展示新闻列表
 */
abstract class NewsListFragment : XFragment(), SwipeRefreshLayout.OnRefreshListener,
        NewsDataSource.LoadNewsListCallback, NewsDataSource.MarkNewsesReadedCallback {
    private var mListAdapter: RecyclerView.Adapter<*>? = null

    protected var mNewsRepository: NewsDataSource? = null
    private var mNewsList: List<News>? = null

    private lateinit var mNewsListFragmentBinding: NewslistFragmentBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mNewsRepository = Injection.provideNewsRepository()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mNewsListFragmentBinding = DataBindingUtil.inflate(inflater, R.layout.newslist_fragment, container, false)

        mNewsListFragmentBinding.newslistSwiperefreshlayoutPullrefresh.setOnRefreshListener(this)
        mNewsListFragmentBinding.newslistRecyclerviewList.layoutManager = LinearLayoutManager(context)
        mNewsListFragmentBinding.newslistRecyclerviewList.addItemDecoration(RecyclerViewItemDecoration(6))

        return mNewsListFragmentBinding.root
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
        refreshNewsList()
    }

    override fun onStop() {
        EventBus.getDefault().unregister(this)
        super.onStop()
    }


    override fun onRefresh() {
        refreshNewsList()
    }

    override fun onNewsLoaded(newsList: List<News>) {
        if (mIsForeground) {
            mNewsList = newsList
            mNewsListFragmentBinding.newslistSwiperefreshlayoutPullrefresh.isRefreshing = false
            showBigCardsList()
            Snackbar.make(mNewsListFragmentBinding.root, "Fetch " + newsList.size + " newses ...", Snackbar.LENGTH_SHORT).show()
        }
    }

    override fun onDataNotAvaiable() {
        Snackbar.make(mNewsListFragmentBinding.root, "No newses refresh...", Snackbar.LENGTH_SHORT).show()
    }

    private fun refreshNewsList() {
        Snackbar.make(mNewsListFragmentBinding.root, "Fetch more newses ...", Snackbar.LENGTH_SHORT).show()
        mNewsListFragmentBinding.newslistSwiperefreshlayoutPullrefresh.isRefreshing = true
        getNewsList()
    }

    abstract fun getNewsList()

    @Subscribe
    fun onShowBigCardsListEvent(event: NewsListShowBigCardsEvent) {
        //FIXME 用双引号是不是有点激进
        showBigCardsList()
    }

    @Subscribe
    fun onShowMinimalListEvent(event: NewsListShowMinimalEvent) {
        //FIXME 用双引号是不是有点激进
        mListAdapter = MinimalAdapter(context!!, mNewsList!!)
        mNewsListFragmentBinding.newslistRecyclerviewList.adapter = mListAdapter
    }

    @Subscribe
    fun onShowCompactListEvent(event: NewsListShowCompactEvent) {
        //FIXME 用双引号是不是有点激进
        mListAdapter = CompactAdapter(context!!, mNewsList!!)
        mNewsListFragmentBinding.newslistRecyclerviewList.adapter = mListAdapter
    }

    @Subscribe
    fun onMarkAsReadEvent(event: NewsListAsReadEvent) {
        //将现有文章置为可读
        val newsIdList = ArrayList<String>()
        for (news in mNewsList!!) {
            newsIdList.add(news.newId!!)
        }
        mNewsRepository?.markNewsesReadedByNewsId(newsIdList, this)
    }

    override fun onMarkNewsesReadedSuccess() {
        //刷新列表
        refreshNewsList()
    }

    override fun onMarkNewsesReadedFaild() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    fun showBigCardsList() {
        mListAdapter = BigCardsAdapter(context!!, mNewsList!!)
        mNewsListFragmentBinding.newslistRecyclerviewList.adapter = mListAdapter
    }
}